### ISSUE TITLE
Cursor disappears on matching symbols in vim

### ISSUE THREAD

**Issue by Skallwar** [jwilm/alacritty#2398]
Hi,
 I'm using vim with molokai colorscheme and when my cursor is on a matching symbol ( like : {} or () ) the cursor block goes black . With kitty, this is not happening since the cursor stays the same. It's really annoying because I tend to "lose" the cursor...

I'm on Arch linux using x11

**Comment by chrisduerr** [comment]
Are you saying that the cursor block goes to the inverse of the cell color but you want it to stay a fixed color?

If so, there's an option in the config to set that.

**Comment by Skallwar** [comment]
For example when I move the cursor over an orange '{' the cursor goes black (or transparent as the background is black) but the '{' is visible.

See the difference:
![cursor](https://user-images.githubusercontent.com/15075443/57065927-ca992480-6cc2-11e9-89f3-a3ece5da1f72.png)

**Comment by chrisduerr** [comment]
Does this happen in vim (not neovim) and is this reproducible with the default configuration?

Also is `showmatch` set or not? And does toggling it change anything?

And which `$TERM` are you using?

**Comment by Skallwar** [comment]
This happens in vim. With default configuration, cursor stays as it should be on matching symbols. You can see all my vimrc in previous images. What surprise me is that with the same vimrc, alacritty reacts differently to kitty 

`showmatch` was not set, but after set it, the problem does not fix.

`$TERM` is not set in my bashrc nor alacritty config. I've try xterm-256color but it doesn't works neither

**Comment by lnikkila** [comment]
I’m also experiencing this, here’s my environment:
- macOS Mojave 10.14.3
- $TERM: `screen-256color`
- NVIM v0.3.4
- VIM - Vi IMproved 8.0 (2016 Sep 12, compiled Aug 17 2018 17:24:51)

Happens in both Vim and Neovim, with default configuration (`vim --clean`). `:set [no]showmatch` doesn’t seem to change anything. Kitty seems to work as described above.

**Comment by chrisduerr** [comment]
> Happens in both Vim and Neovim

Now that's interesting. I'm using NeoVim and am not able to reproduce at all.

Can someone confirm that this was not a problem with 0.3.0?

**Comment by nixpulvis** [comment]
@lnikkila the `:set [no]showmatch` controls a temporary jump to the matching character (see `:h showmatch`), which I believe is different than the permanent style change (see `:h pi_paren.txt`) which is what I'm assuming this issue is talking about.

You can set the value of `MatchParen` at will, I'd be curious what the output of people's `:hi MatchParen` to be.

**Comment by Skallwar** [comment]
@nixpulvis Here is my values for MatchParen: 
`MatchParen     xxx term=reverse cterm=bold ctermfg=233 ctermbg=208 gui=bold guifg=#000000 guibg=#FD971F`

**Comment by nixpulvis** [comment]
Interestingly I've found that when I run this with `kitty` the cursor blinks, could with be related to #2365? I've tried disabling blinking (within `kitty`, we don't yet support it) in Vim with `:set guicursor+=a:blinkon0` as described [here](https://vim.fandom.com/wiki/Configuring_the_cursor), but it doesn't seem to work for me.

Copying your settings for `MatchParen` works for me in both Alacritty and kitty, however kitty seems to ignore the `reverse` and only indicates the cursor position by blinking, while Alacritty displays the cursor as an inverse of the colors described, and doesn't blink (as noted).

**Comment by Skallwar** [comment]
I should note that I use alacritty for 3 month and this "bug" started around 5 days ago without doing anything to any config file, nor updates. It's really weird.

**Comment by chrisduerr** [comment]
Which version are you using? If you downgrade to 0.3.0 does that fix it? Have there maybe been any neovim/vim updates?

**Comment by Skallwar** [comment]
I'm using alacritty 0.3.2-1. Vim has been updated recently but after downgrading vim to a one month old version, the problem is still here.

I have been able to reproduce it on a friend pc just by setting `colorscheme molokai`. I should note that this colorscheme come from `flazz/vim-colorschemes` repo

**Comment by chrisduerr** [comment]
Have you tried downgrading Alacritty to 0.3.0?

**Comment by Skallwar** [comment]
Yes, not working

**Comment by chrisduerr** [comment]
Okay, that's good. That means that the cursor rework was not responsible for this and it's likely not a regression.

I'll see if I can repro in a bit, though it's possible that the colorscheme is at fault here.

**Comment by GuillaumeRochat** [comment]
I've also been losing my cursor in Neovim (0.3.5) since I updated alacritty last week (0.3.2). The cursor comes back when I change the text size with Ctrl+ and Ctrl-.

I'll try to pay more attention to what actions I'm doing to cause that today and update here accordingly.

**Comment by chrisduerr** [comment]
That is a separate issue which should be fixed on the latest master.

**Comment by Skallwar** [comment]
It might be the colorscheme which is at fault, but it is strange that other terminal emulator handled this differently. I just try terminator and again, no problem. As far as I understand, it should be alacritty related

**Comment by iicurtis** [comment]
I am also having a strange issue similar to this. My cursor disappears in insert mode in neovim 0.3.5 but only while in tmux and only after some event happens. I'm not yet sure what the event is that triggers it.

It is reproducible with the following nvim config:

```
set termguicolors
colorscheme desert
```
My alacritty env is set to `xterm-256color` and tmux has the following config:
```
set -s default-terminal "tmux-256color"
set -as terminal-overrides ",alacritty:Tc,xterm-256color:Tc"
```

Strangely enough, opening a new terminal and attaching to the same tmux session works. So the cursor is visible in one window but not the other.

**Comment by dbarrosop** [comment]
I noticed a similar but slightly different issue:

- In alacritty 0.3.0 the cursor always has white background and black foreground
- In alacritty 0.3.2, in normal mode, the cursor's background matches the foreground color. So if I am in a highlighted text (search) it will become yellow, if I am in a symbol it will match the color of that symbol. In insert/replace mode the cursor just disappears.

I personally don't mind one behavior or the other, but the cursor disappearing in insert/repkace mode is certainly a problem

**Comment by JonnyHaystack** [comment]
I've been noticing this same issue since the recent new release. Even more annoyingly, when I am insert mode and it is using a beam cursor instead of block cursor, after a short time this beam cursor becomes completely invisible and I'll never see it again until I restart alacritty. So effectively I'm never able to see the cursor in insert mode in vim which is a big problem for me.

By the way this is with neovim.

**Comment by jonhoo** [comment]
I see this manifest the same way as @iicurtis -- it happens only after I've spent some time with neovim open in tmux in Alacritty. I'm not sure what the triggering event is. I believe what happens is that the cursor eventually stops respecting cursor color, and instead just uses the default of reversing the fg/bg color of the current cursor location. If I open a second alacritty window and open exactly the same thing, the cursor is normal again.

**Comment by Skallwar** [comment]
In vim I have noticed sometimes a slight delay for this problem to appear, but only with small file like 50 lines. With bigger filer it appears immediately.

> ![cursor](https://user-images.githubusercontent.com/15075443/57065927-ca992480-6cc2-11e9-89f3-a3ece5da1f72.png)
>

Basically, this seems to be a background/foreground inversion. In the picture we can see that the curly bracket should be black instead of orange and the background should be orange instead of black

**Comment by nixpulvis** [comment]
@Skallwar you've linked the same image as before, was that intentional?

Also, @Skallwar, you mention that it only starts after a delay? Can you record a gif or something of this please, I'm not quite sure what's going on specifically with your case yet still.

After my initial investigation here, it seemed like it could be related to the blinking cursor and how we handle reversing text, but I'm not even 100% sure what you think correct behavior here is.

**Comment by Skallwar** [comment]
@nixpulvis Yes, it was intentional just as a reminder to prove my point without scrolling to the top of the issue.

The delay is quite random I'm not sure if I will be able to record anything but I will try. 

About the correct behavior, for me it should be what kitty is doing which in the picture is: exact same color as the matching symbol (which is the inverse of what alacritty is doing)
I have also tried to deactivated cursor blinking in kitty and the color of the cursor on a matching symbol is as follow:
![cursor](https://user-images.githubusercontent.com/15075443/57491249-f6646d80-72b3-11e9-8a83-cbd6ec555e30.png)

However, since terminator is doing the same thing as alacritty, I don't know what to think anymore...
In my opinion, kitty's approach is the best one right now, mainly because I keep loosing my cursor since it is not blinking.

**Comment by nixpulvis** [comment]
@Skallwar so, the question for you is, would you consider fixing #302 a solution to this?

**Comment by Skallwar** [comment]
@nixpulvis I see 3 solutions here (ranked from best to worst in my opinion):
- Fix #302 which is a nice feature to have
- Try to always show the same colors than the matching symbol (like kitty does in the first image I sent).
- Fix cursor color to black & white in the config (see #704). It is a quick fix to the problem, but losing automatic color adaptation is quite sad, since it is pretty

**Comment by Skallwar** [comment]
> Interestingly I've found that when I run this with `kitty` the cursor blinks, could with be related to #2365? I've tried disabling blinking (within `kitty`, we don't yet support it) in Vim with `:set guicursor+=a:blinkon0` as described [here](https://vim.fandom.com/wiki/Configuring_the_cursor), but it doesn't seem to work for me.
> 
In kitty you can disable blinking by setting `cursor_blink_interval 0` in `~/.config/kitty/kitty.conf`

> Copying your settings for `MatchParen` works for me in both Alacritty and kitty, however kitty seems to ignore the `reverse`
>
Maybe ignore `reverse` it what will fix : 
>     * Try to always show the same colors than the matching symbol (like kitty does in the first image I sent).
>

**Comment by trescenzi** [comment]
Is there any movement here? I'm having the same issue as @iicurtis and while it's not tough to disconnect from the tmux session and then reconnect its definitively reducing quality of life. Thanks!

**Comment by dumbbell** [comment]
Hi! I also see the cursor in NeoVim disappear in insert mode, dafter some time or event. It never disappears in command mode. I don't think I changed the cursor configuration, so it's a block cursor in command mode and a "pipe" cursor in insert mode.

I don't use tmux in this context, so I believe it's unrelated to tmux.

**Comment by noahfrederick** [comment]
I'm not sure we're all talking about the same thing, but the issue that I was having related to this has been resolved by updating Neovim from 0.3.5 to 0.3.7.

## Symptoms

- Cursor color did not respect Alacritty's cursor color settings, but would instead use reverse (in Neovim only)
- Cursor was never visible in insert mode (should show bar cursor)
- Cursor was sometime invisible in normal mode (e.g., MatchParen)

I did not experience any of this after a delay as others have reported. The symptoms were immediate and persistent.

## Context

- Issue occurred in Neovim 0.3.5, under tmux 2.8 and 2.9, in Alacritty 3.2. Didn't test other programs or configurations.
- Neovim colorscheme defines a cursor color (black on white) (I believe this is ignored in the TUI)
- Static cursor color (black on white) is set in alacritty.yml

**Comment by kchibisov** [comment]
@Skallwar I've checked xterm/gnome-terminal and they behave exactly the same as alacritty. It happens because our configuration sets cursor to reverse by default(xterm and gnome-terminal also do this), so setting cursor to a fixed value in a config should resolve your issue (you've mentioned that, but I don't see other option, since xterm works the same). Kitty doesn't have an configuration level reverse option as I can see, so that's why it works for you.

In case you still think that it's our fault, please provide the whole vim + alacritty configuration, since I can't repro your issue on the latest master + your `colorscheme`.

FWIW we're ignoring reversing a cursor with escape.

For other people on `v0.3.2` It seems like your issue was related to an old cursor rework that was fixed in `v0.3.3` which could cause cursor disappearing.

**Comment by Neurognostic** [comment]
I was having a similar issue on v0.3.3 where the cursor and background color would be the same and had the affect of making the cursor "disappear". Except in my case, it would only occur when returning to normal mode from insert mode when the cursor was in a whitespace field.

The environment stack where this would occur for me was `alacritty.exe -> zsh -> tmux -> nvim`.
It would not occur without tmux and it would not occur with tmux+vim.

I am happy to report I am no longer seeing this issue after upgrading Alacritty to v4.0 and can continue to enjoy Alacritty's neat cursor colors.

**Comment by Skallwar** [comment]
I'm getting quite confuse. I don't know what is the "correct" behavior and if the issue is related to the colorscheme or alacritty itself. I think some people here have another issue due to tmux (maybe start an other one).

> @Skallwar I've checked xterm/gnome-terminal, and they behave exactly the same as alacritty. It happens because our configuration sets cursor to reverse by default(xterm and gnome-terminal also do this), so setting cursor to a fixed value in a config should resolve your issue
>
Yep I will accept this as a solution while waiting for the blinking cursor feature #302

**Comment by Neightro** [comment]
I think I'm experiencing this bug; I'm on Manjaro Linux with Alacritty 0.6.0, Xorg 1.20.10. I have a cursor color set in my alacritty.yml, but as soon as I place the cursor over anything formatted with the 'reverse' style the cursor turns into the background color.

It's nothing specific to Vim; I can reproduce it just by selecting over my cursor with my mouse; this method works for me if I'm in a shell (tried Zsh and Bash), but strangely not in Vim.

![image](https://user-images.githubusercontent.com/23202923/101971520-b7607180-3be6-11eb-98e3-ea611716e2f2.png)

Alternatively, in Vim/Neovim, I can run the command `:hi Function cterm=reverse gui=reverse` and hover over a function name; same issue arises. Playing with different highlighting groups might help; `Function` and `Comment` both work for different color schemes for me. The screenshot here is taken with two-firewatch.

![image](https://user-images.githubusercontent.com/23202923/101971796-c3e5c980-3be8-11eb-8cb3-a622182ec053.png)

**Comment by kchibisov** [comment]
We have an explicit logic to speculate with cursor colors if the contrast is low(e.g. you won't see it at all). So on a screenshots you've linked cursor is visible, just not the colors from your config, since you won't see it if we keep using the colors from config.

**Comment by Neightro** [comment]
@kchibisov Is there a way to disable this logic? I don't think that it makes the cursor easier to see, to be honest; most of time time I find it does the opposite.

**Comment by kchibisov** [comment]
No, How could you see a cursor which would have he same fixed color as your background? You can't see `#ffffff` on `#ffffff` for example. if it's easier for you to see `#ffffff` on `#ffffff` tell me how to do that .

**Comment by Neightro** [comment]
I should have added; my cursor is orange (#e19b14); I should be able to see more than fine against a white background.

**Comment by kchibisov** [comment]
Maybe then it's some other logic, since on a screenshot your cursor isn't orange at all. In that case share your config.

**Comment by Neightro** [comment]
Sure thing!

[alacritty.yml.txt](https://github.com/alacritty/alacritty/files/5683408/alacritty.yml.txt)

**Comment by kchibisov** [comment]
it's indeed due to contrast logic, please open a new issue with your config and describe that you get cursor inverted when it likely shouldn't. thx.

**Comment by AckslD** [comment]
I'm just trying out alacritty and having this issue in neovim using the gruvbox colorscheme from [here](https://github.com/alacritty/alacritty/wiki/Color-schemes). See gif below for difference with kitty. Anyone found a fix for this?

![alacritty](https://user-images.githubusercontent.com/23341710/124096403-d62b3600-da5a-11eb-86c8-eb914d17b158.gif)

**Comment by devnoot** [comment]
I'm actually having this same issue using a custom colorscheme here

```

# google dark color theme
colors:
  primary:
    background: '0x202124'
    foreground: '0xc5c8c6'
  normal:
    black:   '0x202124'
    red:     '0xcc342b'
    green:   '0x34A853'
    yellow:  '0xFBBC04'
    blue:    '0x4285F4'
    magenta: '0xa36ac7'
    cyan:    '0x3971ed'
    white:   '0xffffff'
  bright:
    black:   '0x969896'
    red:     '0xcc342b'
    green:   '0x34A853'
    yellow:  '0xFBBC04'
    blue:    '0x4285F4'
    magenta: '0xa36ac7'
    cyan:    '0x3971ed'
    white:   '0xffffff'
```

**Comment by karlyeurl** [comment]
Hello,

I have the exact same problem as the OP ([this one](https://user-images.githubusercontent.com/15075443/57065927-ca992480-6cc2-11e9-89f3-a3ece5da1f72.png)). I see this bug is closed, and yet it still appears.

Switching colorschemes makes the behavior exceedingly inconsistent, and the lack of background on the cursor when navigating code and hitting brackets/parens makes Alacritty a lesser options compared to other terminal emulators.

Has anyone found a suitable way around this? It's kind of my last hurdle to overcome before fully switching to Alacritty.

**Comment by kchibisov** [comment]
You can set the cursor color in the config to some fixed value if you don't want it to disappear due to inversion?

**Comment by karlyeurl** [comment]
Yes, that can be achieved, but I am reluctant to do so for two reasons:
- every time I choose a new theme, I will have to alter that fixed value
- no other terminal emulator behaves like that (xterm, termite, urxvt, kitty), which leads me to believe the issue stems from alacritty, not from vim or the colorschemes

**Comment by felixmde** [comment]
I can also still observe this behavior:

![image](https://github.com/alacritty/alacritty/assets/59488930/b5054c14-7f91-414b-bb25-97f7bda097c9)

My current workaround is to set `highlight MatchParen ctermfg=NONE` manually which gives me this:

![image](https://github.com/alacritty/alacritty/assets/59488930/a68154ee-9fae-4e37-bfe0-5ea834af6af7)

**Comment by kchibisov** [comment]
I'd assume that in st and in alacritty you do inverse the cursor? Because it clearly seems like alacritty inverses cursor, but `st` not.

**Comment by felixmde** [comment]
> I'd assume that in st and in alacritty you do inverse the cursor? Because it clearly seems like alacritty inverses cursor, but `st` not.

I am not sure if I understand what inverse means here. If it means cursor shape 'block', both alacritty and st use 'block'. Interestingly, when I change alacritty to 'underline', MatchParen on the opening parentheses becomes visible (because there is only the bar under the cursor).

I know that MatchParen term also has a reverse option, but they are configured the same:

```
Alacritty:
MatchParen     xxx term=reverse ctermfg=234 ctermbg=241 guifg=#19171c guibg=#655f6d
st:
MatchParen     xxx term=reverse ctermfg=234 ctermbg=241 guifg=#19171c guibg=#655f6d
```

Also, apologies for the initial screenshot. I noticed that st was not actually in focus, so I have updated that.

Lastly, when alacritty goes out of focus, the cursor appears for the opening parentheses.

I don't know if that is enough data to come to any reasonable conclusions, but please let me know if anyone has any further suggestions.

**Comment by kchibisov** [comment]
The default cursor in alacritty changes the colors depending on the background.

`term=reverse` will reverse the colors as well, so by doing so twice you'll get no-reverse, meaning it'll match your background and foreground, so you feel like it's invisible.

You have 2 options on how to solve it:

1. Don't reverse the cursor in your vim scheme, simply remove the `reverse` and swap `bg` and `fg` values yourself.
2. Set cursor colors in alacritty config to some fixed values, so reverse logic won't touch it.

There's no bug here. The only thing we could do is to not reverse the reverse cursor if the cell is already `reverse`, do we want to do something like that @chrisduerr ?

**Comment by chrisduerr** [comment]
> do we want to do something like that @chrisduerr ?

I don't.

**Comment by felixmde** [comment]
> Don't reverse the cursor in your vim scheme, simply remove the reverse and swap bg and fg values yourself.

That worked. Thanks.

So, ultimately, this is an issue with the color scheme (I guess).
